pkg/npm/http-api/src/index.ts

```diff
@@ -0,0 +1,455 @@
+import { Action } from '../../api';
+
+import { AuthenticationInterface, SubscriptionInterface, CustomEventHandler, PokeInterface, SubscriptionRequestInterface, headers, UrbitInterface, SSEOptions } from './types';
+import UrbitApp from './app/base';
+import { uncamelize, hexString } from './utils';
+
+/**
+ * A class for interacting with an urbit ship, given its URL and code
+ */
+export class Urbit implements UrbitInterface {
+  /**
+   * UID will be used for the channel: The current unix time plus a random hex string
+   */
+  uid: string = `${Math.floor(Date.now() / 1000)}-${hexString(6)}`;
+
+  /**
+   * Last Event ID is an auto-updated index of which events have been sent over this channel
+   */
+  lastEventId: number = 0;
+
+  lastAcknowledgedEventId: number = 0;
+
+  /**
+   * SSE Client is null for now; we don't want to start polling until it the channel exists
+   */
+  sseClient: EventSource | null = null;
+
+  /**
+   * Cookie gets set when we log in.
+   */
+  cookie?: string | undefined;
+
+  /**
+   * A registry of requestId to successFunc/failureFunc
+   * 
+   * These functions are registered during a +poke and are executed
+   * in the onServerEvent()/onServerError() callbacks. Only one of
+   * the functions will be called, and the outstanding poke will be
+   * removed after calling the success or failure function.
+   */
+
+  outstandingPokes: Map<number, object> = new Map();
+
+  /**
+   * A registry of requestId to subscription functions.
+   * 
+   * These functions are registered during a +subscribe and are
+   * executed in the onServerEvent()/onServerError() callbacks. The
+   * event function will be called whenever a new piece of data on this
+   * subscription is available, which may be 0, 1, or many times. The
+   * disconnect function may be called exactly once.
+   */
+
+  outstandingSubscriptions: Map<number, SubscriptionInterface> = new Map();
+
+  /**
+   * Ship can be set, in which case we can do some magic stuff like send chats
+   */
+  ship?: string | null;
+
+  /**
+   * If verbose, logs output eagerly.
+   */
+  verbose?: boolean;
+
+  /**
+   * All registered apps, keyed by name
+   */
+  static apps: Map<string, typeof UrbitApp> = new Map();
+
+  /** This is basic interpolation to get the channel URL of an instantiated Urbit connection. */
+  get channelUrl(): string {
+    return `${this.url}/~/channel/${this.uid}`;
+  }
+
+  get fetchOptions(): any {
+    const headers: headers = {
+      'Content-Type': 'application/json',
+    };
+    if (!isBrowser) {
+      headers.Cookie = this.cookie;
+    }
+    return {
+      credentials: 'include',
+      headers
+    };
+  }
+
+  /**
+   * Constructs a new Urbit connection.
+   *
+   * @param url  The URL (with protocol and port) of the ship to be accessed
+   * @param code The access code for the ship at that address
+   */
+  constructor(
+    public url: string,
+    public code: string
+  ) {
+    return this;
```

Comment:
double return??

Reply:
I had put in the proxy technique but it's ugly and of questionable usefulness. I left the old code in in case I found a reason to go back.